### PR TITLE
1622 stepper does not evaluate arguments when it should

### PR DESCRIPTION
**Pull request by hanscau**
Fix the issue by rearranging the reduction steps of function call

Refactor code for easier readability 

Added comments

Added tests

### CODE REVIEW

File: src/stepper/stepper.ts [hunk 1]
```diff
@@ -1683,6 +1683,29 @@ function reduceMain(
...
+          ]
+        } else if (
+          currentArg.type === 'Identifier' &&
+          !(currentArg.name in context.runtime.environments[0].head)
+        ) {
+          throw new errors.UndefinedVariable(currentArg.name, currentArg)
+        }
+      }
+
+      //Reduce callee until it is irreducible
```
Comment (martin-henz) [inline src/stepper/stepper.ts:1708]: Should we not reduce the callee, before the arguments?
Example:
```
((x => x)(x => x))(1 + 2);
```
currently evaluates the callee first, and then 1 + 2. That is the correct order.

File: src/stepper/stepper.ts [hunk 2]
```diff
@@ -1683,6 +1683,29 @@ function reduceMain(
...
+            ast.callExpression(callee as es.Expression, reducedArgs as es.Expression[], node.loc),
+            cont,
+            path,
+            str
+          ]
+        } else if (
+          currentArg.type === 'Identifier' &&
+          !(currentArg.name in context.runtime.environments[0].head)
+        ) {
+          throw new errors.UndefinedVariable(currentArg.name, currentArg)
```
Comment (martin-henz) [inline src/stepper/stepper.ts:1704]: Should this be UnassignedVariable? We shouldn't ahve UndefinedVariable anywhere, because we use the validator.

**Review by martin-henz — CHANGES_REQUESTED** [review]

File: src/stepper/stepper.ts (hunk 2 above)
Comment (hanscau) [reply src/stepper/stepper.ts:1704]: Yeap, it is redundant. Will remove it

**Review by martin-henz — APPROVED** [review]
Thanks for fixing the issues.

**Pull request merged**